### ISSUE TITLE
How to gather completions before computing rewards in GRPOTrainer

### ISSUE THREAD

**Issue by rubickkcibur** [huggingface/trl#3896]
Hi, 
I found that the `reward_funcs` passed to GRPOTrainer is used per-device.
That is, if I set `num_generation=16`, `per_device_train_batch_size=4`, my customized reward function can only receive `4` completions.
However, my customized reward function calculates rewards depending on a global view over all `16` completions for each question.
How can I implement this?

**Comment by LeonEricsson** [comment]
Are you running with some kind of data parallelism? If not, you should be seeing all completions in your reward function. 

<details>
<summary> see code example that works</summary>

```python
import random

from datasets import load_dataset
from transformers import AutoModelForCausalLM

from trl import GRPOConfig, GRPOTrainer

dataset = load_dataset("trl-lib/tldr", split="train[:50]")

def reward_random(completions, **kwargs):
    assert len(completions) == 8 # == num_generations; this runs fine
    return [random.random() for _ in range(len(completions))]

model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct", torch_dtype="auto", device_map="auto")

trainer = GRPOTrainer(
    model=model,
    reward_funcs=reward_random,
    train_dataset=dataset,
    args=GRPOConfig(
        steps_per_generation=2, # must be set 2 two because the generation_batch_size must be evenly divisible by num_generations
        per_device_train_batch_size=4,
        num_generations=8,
        max_completion_length=128,
        logging_steps=1,
        max_steps=1,
        num_train_epochs=1,
        report_to="none",
        bf16=True,
    ),
)

trainer.train()
```
</details>

if you are running across multiple devices you won't have a global view of completions as they are distributed across devices to increase efficiency.

**Comment by rubickkcibur** [comment]
Yes, I am running on 8 GPUs, using accelerate with Deepspeed Zero3. I looked into the implementation of GRPOTrainer and found your saying. The completions are distributed across devices and gathered after reward computation. 
My need is just to gather the distributed completions first and then compute my custom reward over the gathered completions.
Maybe I need to overload the related functions in a custom GRPOTrainer?

**Comment by LeonEricsson** [comment]
yes, this would require rewriting the [_calculate_rewards](https://github.com/huggingface/trl/blob/12fc85fd13ac79247542db588f55137f1cc5a8f2/trl/trainer/grpo_trainer.py#L999) to gather before reward calculation.

**Comment by rubickkcibur** [comment]
I see. Thanks for your response!
